src/content/blog/cheerp-typescript.mdx

```diff
@@ -0,0 +1,393 @@
+
+```cpp title="square.cpp"
+[[cheerp::jsexport]]
+double square(double x) {
+    return x * x;
+}
+```
+
+```js title="index.js"
+import initSquare from "./square.js";
+const { square } = await initSquare();
+console.log(square(2)); // prints "4"
+```
+
+### Type checking
+
+Note that the `square` function in C++ takes an argument of type `double`.
+When calling this function from C++, the compiler would ensure at compile time
+that we cannot pass, for example, a string to the `square` function. Let's see
+what happens when we try to pass in a string from JavaScript.
+
+```js title="index.js"
+console.log(square("hello")); // prints "NaN"
+```
+
+JavaScript will not give us a compile error, not even a runtime error, we only
+get `NaN`. Mistakes like this can be easy to make, and it would be nice if the
+compiler could help us out here.
+
+This is where TypeScript comes in. When given a declaration for the `square`
+function, the TypeScript compiler can, at compile time, check that the
+arguments have the correct types. Let's start by manually writing a declaration
+file.
+
+```ts title="square.d.ts"
+export default function(): Promise<{
+    square(x: number): number;
+}>;
+```
+
+Now, when we compile our code, the TypeScript compiler will yell at us if we
+try to pass a string to the `square` function.
+
+```
+index.ts:3:20 - error TS2345: Argument of type 'string' is not assignable to parameter of type 'number'.
+
+3 console.log(square("hello"));
+                     ~~~~~~~
+```
+
+### Automation
+
+Writing these declaration files by hand is a pain. We already wrote down the
+types once in C++, why do we have to repeat the same function signature again?
+
+As you have probably guessed, Cheerp can now automatically generate TypeScript
+declaration files for us. Simply add the `-cheerp-make-dts` flag.
+
+```sh "-cheerp-make-dts"
+/opt/cheerp/bin/clang++ square.cpp \
+    -o square.js \
+    -target cheerp \
+    -cheerp-make-module=es6 \
+    -cheerp-make-dts
+```
+
+After running this command, you should find a `square.d.ts` with content
+similar to what we wrote before.
+
+### Complex types
+
+The TypeScript declaration generator is also able to translate your own
+`[[cheerp::jsexport]]` types and types from Clientlib.
+
+```cpp title="complex.cpp"
+class [[cheerp::jsexport]] Vec2 {
+    double x, y;
+public:
+    Vec2(double x, double y) : x(x), y(y) {}
+    double length() const { return std::sqrt(x * x + y * y); }
+};
+
+[[cheerp::jsexport]]
+client::TArray<client::String*>* keys(client::Object* object) {
+    return client::Object::keys(object);
+}
+```
+
+```ts title="complex.d.ts"
+export interface Vec2 {
+        length(): number;
+        delete(): void;
+}
+type __Options = { buffer: ArrayBuffer } | { absPath: String | URL };
+export default function(options?: __Options): Promise<{
+        Vec2: {
+                new(x: number, y: number): Vec2;
+        };
+        keys(_object: object): Array<string>;
+}>;
+```
+
+### Internals
```

Comment:
I do like the existence of this section, although maybe it'd be better off as an expandable section, an "explanation" page on the docs, or a separate blog post? Then we link to it. Alternatively we move it to the end of the post if possible

Some readers will like it, but it's also long enough to turn others off.

Reply:
After reading the entire post again, I think for some readers the internals section will actually be the most interesting part, and I think it's good to leave it in as is.